### PR TITLE
ENYO-6200: Add position prop to ProgressBarTooltip

### PR DESCRIPTION
**Pull request by ryanjduffy**
### Checklist

* [X] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [X] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [X] Documentation was added or is not needed

* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Authors are unable to control the position of `ProgressBarTooltip` and `SliderTooltip` effectively with the current API. For example, it is not possible to force the tooltip to always display to one side.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
* Add `position` prop with broader set of values to replace `side`
* Deprecate `side`

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)
* Reordered knobs in `ProgressBar` and `Slider` stories to force the component tab to appear before the tooltip tab
* Added an internal `deprecated` prop type to simplify deprecating props
* Added a custom prop type validate for `position` to warn when using invalid values with `orientation`

### CODE REVIEW

**Comment by ryanjduffy** [comment]
Oof. Should add some unit tests to this.

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 1]
```diff
@@ -11,22 +12,84 @@ import Tooltip from '../TooltipDecorator/Tooltip';
...
-	return valid ? side : 'before';
+		if (orientation === 'horizontal') {
+			switch (position) {
+				case 'above':
+				case 'below':
+					return ['auto', position];
+				case 'after':
+				case 'before':
+				case 'left':
+				case 'right':
```
Comment (webOS101) [inline packages/moonstone/ProgressBar/ProgressBarTooltip.js]: As I mentioned before, I super-prefer to see the defaults in `default:`.  I guess there are more than twice as many valid ones as these 4 invalid ones.  Could possibly get all the 'above' and 'below' options on one line each?

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 2]
```diff
@@ -97,10 +213,18 @@ const ProgressBarTooltipBase = kind({
 		 * * `'right'` renders to the right of a `vertical` ProgressBar/Slider regardless of locale
 		 *
 		 * @type {String}
-		 * @default 'before'
+		 * @deprecated Deprecated since 3.1 until 4.0 in favor of `position`
```
Comment (webOS101) [inline packages/moonstone/ProgressBar/ProgressBarTooltip.js]: I think this duplicates deprecated.  Will verify.

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 3]
```diff
@@ -135,39 +258,54 @@ const ProgressBarTooltipBase = kind({
...
+		arrowAnchor: ({orientation, position, proportion, rtl}) => {
 			if (orientation === 'vertical') return 'middle';
-			return proportion > 0.5 ? 'left' : 'right';
+
+			const [h] = getSide(orientation, false, position);
+			switch (h) {
+				case 'auto':
+					return proportion > 0.5 ? 'left' : 'right';
+				case 'before':
+					return rtl ? 'right' : 'left';
```
Comment (webOS101) [inline packages/moonstone/ProgressBar/ProgressBarTooltip.js]: We'll need to communicate this new RTL ability in the release notes.  It should probably also be called out above in the description (at least, when we suppress the default behavior).

File: packages/sampler/stories/default/ProgressBar.js [hunk 4]
```diff
@@ -15,20 +15,25 @@ storiesOf('Moonstone', module)
 	.add(
 		'ProgressBar',
 		() => {
-			const side = select('side', ['after', 'before', 'left', 'right'], ProgressBarTooltipConfig, 'before');
+			// added here to force Storybook to put the ProgressBar tab first
+			const disabled = boolean('disabled', ProgressBarConfig);
```
Comment (webOS101) [inline packages/sampler/stories/default/ProgressBar.js]: This is a bummer we need to do this but I guess it makes sense.  We can fix other stories (if any) later that have a similar problem

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js (hunk 2 above)
Comment (webOS101) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: Nope, this is fine!

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js (hunk 1 above)
Comment (ryanjduffy) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: okay

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 5]
```diff
@@ -65,6 +128,59 @@ const ProgressBarTooltipBase = kind({
 		 */
 		percent: PropTypes.bool,
 
+		/**
+		 * Position of the tooltip with respect to the progress bar.
+		 *
+		 * * For `orientation="horizontal"` progress bars, the default value is `'above'`.
+		 * * For `orientation="vertical"` progress bars, the default value is `'above'`.
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: yep

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js (hunk 3 above)
Comment (ryanjduffy) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: It is vaguely noted but agree it should be expanded.

File: packages/moonstone/ProgressBar/ProgressBarTooltip.js [hunk 6]
```diff
@@ -11,22 +12,84 @@ import Tooltip from '../TooltipDecorator/Tooltip';
 
 import css from './ProgressBarTooltip.module.less';
 
+// prop-type validator that warns on invalid orientation + position
+/* istanbul ignore next */
+const validatePosition = (base) => (props, key, componentName, location, propFullName, ...rest) => {
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: Fair point

[the remark this answers is not in the document]
Comment (webOS101) [reply packages/moonstone/ProgressBar/ProgressBarTooltip.js]: To clarify my point, I meant only in propTypes we don't cross-validate.  We do cross-validate in other spots.

**Pull request merged**